### ISSUE TITLE
puppet service discovery mechanism

### ISSUE THREAD

**Issue by rismoney** [VictoriaMetrics/VictoriaMetrics#5744]
### Is your feature request related to a problem? Please describe

new feature request

### Describe the solution you'd like

prometheus currently has the ability to perform service discovery by way of the puppetdb.
This code is here- https://github.com/prometheus/prometheus/blob/main/discovery/puppetdb/puppetdb.go

### Describe alternatives you've considered

We are currently using prometheus and are exploring the ability to migrate to victoria, but our workflow is heavily tied to package resources, tagging, and other metadata that comes via puppet.

**Comment by hagen1778** [comment]
Hello @jiekun !

Thanks for your interest and PR for OVH! We didn't plan to add Puppetdb because there was no demand for it. As you see, this ticket doesn't have upvotes and I haven't seen demand in other community channels as well. 

But if you'd like to contribute puppetdb SD that would be great and will improve compatibility with Prometheus.

**Comment by sboyd-m** [comment]
Saying "this ticket doesn't have upvotes" when it has only been open for 6 hours is fairly disingenuous. Consider this an upvote. We have been looking at moving from graphite to VictoriaMetrics and like what we see, and we have an all-puppet environment.

**Comment by jiekun** [comment]
Hi @rismoney @fzuerkeraguilar @sboyd-m, I could provide the Docker image or a binary build of vmagent that supports Puppet service discovery. I wanted to ask if you could assist with testing and verifying if it works for you. Please let me know, thanks!

**Comment by sboyd-m** [comment]
@jiekun We may be able to test, but I think it would be better if we could at least see the contents of a work-in-progress patch, rather than running just a binary.

**Comment by jiekun** [comment]
> @jiekun We may be able to test, but I think it would be better if we could at least see the contents of a work-in-progress patch, rather than running just a binary.

Sure. The Pull request is on the way.

**Comment by jiekun** [comment]
Hi @rismoney @fzuerkeraguilar @sboyd-m I've created a pull request for this feature.

https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7172

In the mean time, you could test this feature by:
- Build it from the `feature/puppet-sd` branch. See: https://docs.victoriametrics.com/#how-to-build-from-sources.
- Request a test image/binary. Please let me know if you need help.

Any feedback is greatly appreciated. Thanks!

**Comment by sboyd-m** [comment]
@jiekun Maybe I am dumb, but victoria metrics seems unable to skip verifying the puppetdb server certificate. I have the following config:

```yaml
---
scrape_configs:
  - job_name: puppetdb_job
    tls_config:
      insecure_skip_verify: true
    puppetdb_sd_configs:
      - url: https://<snip>:8081
        port: 8081
        query: 'resources { type = "Service" and title = "httpd" }'
...
```

However, when starting vm I get `tls: failed to verify certificate: x509: certificate signed by unknown authority`

I also tried to add the path to the puppetca root cert, but that did not fix anything. Does Victoria Metrics honor the `tls_config` section from Prometheus config?

**Comment by jiekun** [comment]
@sboyd-m Hi. Hi. Thank you for cooperating with the testing. I think you can make the configuration changes like this and try again:

```yaml
---
scrape_configs:
  - job_name: puppetdb_job
    puppetdb_sd_configs:
      - url: https://<snip>:8081
        port: 8081
        query: 'resources { type = "Service" and title = "httpd" }'
        tls_config:
          insecure_skip_verify: true
...
```

Also, please build it again since I push 1 bugfix commit this morning: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7172/commits

**Comment by sboyd-m** [comment]
Ahh, that's quite embarrasing, I feel like I had tried that.

After updating the yaml, I was able to have VictoriaMetrics discover a machine from PuppetDB, and it is successfully populating metrics. It seems to function on a basic level.

**Comment by jiekun** [comment]
Thank you @sboyd-m. Please don't hesitate to provide feedback on any issues you encounter in the future. We still need to review it before it can be released in the upcoming version.

**Comment by zekker6** [comment]
Support of puppet service discovery was released in [v1.106.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.106.0). closing feature request as completed.
